google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessage.java

```diff
@@ -0,0 +1,322 @@
+    }
+    if (json.length() == 0) {
+      throw new IllegalArgumentException("JSONObject is empty.");
+    }
+    return convertJsonToProtoMessageImpl(
+        protoSchema, json, "root", /*topLevel=*/ true, allowUnknownFields);
+  }
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @param jsonScope Debugging purposes
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @param topLevel checks if root level has any matching fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  private static DynamicMessage convertJsonToProtoMessageImpl(
+      Descriptor protoSchema,
+      JSONObject json,
+      String jsonScope,
+      boolean topLevel,
+      boolean allowUnknownFields)
+      throws IllegalArgumentException {
+
+    DynamicMessage.Builder protoMsg = DynamicMessage.newBuilder(protoSchema);
+    String[] jsonNames = JSONObject.getNames(json);
+    if (jsonNames == null) {
+      return protoMsg.build();
+    }
+    int matchedFields = 0;
+    for (int i = 0; i < jsonNames.length; i++) {
+      String jsonName = jsonNames[i];
+      // We want lowercase here to support case-insensitive data writes.
+      // The protobuf descriptor that is used is assumed to have all lowercased fields
+      String jsonLowercaseName = jsonName.toLowerCase();
+      String currentScope = jsonScope + "." + jsonName;
+      FieldDescriptor field = protoSchema.findFieldByName(jsonLowercaseName);
+      if (field == null) {
+        if (!allowUnknownFields) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "JSONObject has fields unknown to BigQuery: %s. Set allowUnknownFields to True to allow unknown fields.",
+                  currentScope));
+        } else {
+          continue;
+        }
+      }
+      matchedFields++;
+      if (!field.isRepeated()) {
+        fillField(protoMsg, field, json, jsonName, currentScope, allowUnknownFields);
+      } else {
+        fillRepeatedField(protoMsg, field, json, jsonName, currentScope, allowUnknownFields);
+      }
+    }
+
+    if (matchedFields == 0 && topLevel) {
+      throw new IllegalArgumentException(
+          "There are no matching fields found for the JSONObject and the protocol buffer descriptor.");
+    }
+    DynamicMessage msg;
+    try {
+      msg = protoMsg.build();
+    } catch (UninitializedMessageException e) {
+      String errorMsg = e.getMessage();
+      int idxOfColon = errorMsg.indexOf(":");
+      String missingFieldName = errorMsg.substring(idxOfColon + 2);
+      throw new IllegalArgumentException(
+          String.format(
+              "JSONObject does not have the required field %s.%s.", jsonScope, missingFieldName));
+    }
+    if (topLevel && msg.getSerializedSize() == 0) {
+      throw new IllegalArgumentException("The created protobuf message is empty.");
```

Comment:
out of curiosity, when can we hit this condition? I thought we already have "matchedFields == 0 && topLevel" check above?

Reply:
So if there's only 1 repeated field, and the json passes in an empty array, then it will trigger this condition.